### PR TITLE
feat(v3.6.0): tup 230 make form plugin work

### PR DESCRIPTION
**Pull request by wesleyboar**
## Overview

Form plugin: update, configure, style*, document.

<sub>* The styles only use inspiration from [new TUP login form](https://xd.adobe.com/view/1bebd965-c072-4fac-8727-ba61c5ca3baf-4eb8/screen/462fd224-c85d-45df-9796-9d21eb7d3b1a/specs/). Design will later give standard CMS form styles (and later CMS-specific button styles).</sub>

## Related

- [TUP-230](https://jira.tacc.utexas.edu/browse/TUP-230)
- mimics https://github.com/TACC/Core-CMS/pull/498 but on v3.6.0
- ~~https://github.com/TACC/Core-CMS-Resources/pull/144~~ https://github.com/TACC/Core-CMS-Resources/pull/145

## Changes

- pin core-styles to version compatible with CMS v3.6.0
- install `django-recaptcha`
- replace `djangocms-forms` with `djangocms-forms-maintained`
- isolate and document form plugin settings
- copy `c-button` from core-styles v0.5+
- (minimally) overwrite `c-button`
- style form elements (generic and plugin)
- load jQuery with AJAX (to support form)
- sample code to customize captcha
- override form template to:
    1. load css
    2. fix console error
- add required url entry
- add frontera colors for primary button

## Testing

Deployed on [Frontera Pre-Prod](https://pprd.frontera-portal.tacc.utexas.edu/).
Tests on [TUP-230 test page](https://pprd.frontera-portal.tacc.utexas.edu/test/tup-230-form-plugin/). (Must log in.)

1. Create form plugin instance.
    - Complete _required_ settings (no e-mail).[^1]
    - Check "Save to database".
3. Submit form—see success message.[^2]
4. Submit form again—see success message.[^2]
5. Edit form plugin instance.
    - Add SPAM PROTECTION[^3]
6. Submit form—see success message.[^2]
7. In "Site administration", open "Form submissions".
8. Confirm your submissions are there.

[^1]: E-mail feature is delayed until I have bandwidth to research it and configure or fix it.
[^2]: The success message is top center of screen white on black (a `.cms_message`). These should be styled better.
[^3]: There should only be options "None" and "Recaptcha", chose "Recaptcha"

## Screenshots

<details><summary>Screenshots of Fields & Errors</summary>

| Fields | Errors |
| - | - |
| fields | errors |

</details>

<details><summary>Videos of Form Submission with & w/out Captcha</summary>

https://user-images.githubusercontent.com/62723358/173920493-afcdd29f-fddc-4b02-8ac8-ae31cc76b92f.mov

https://user-images.githubusercontent.com/62723358/173920496-99307343-247b-4760-a1bf-c08238de3b80.mov

</details>

## Notes

Authorized developer should add/update a Recaptcha account with domain/s and add secret key to frontera server.
1. [Confluence: **Google reCAPTCHA**, Analytics & Site Verification](https://confluence.tacc.utexas.edu/x/PQCCB)
3. [TACC/Core-Portal-Deployments Frontera CMS captcha settings](https://github.com/TACC/Core-Portal-Deployments/blob/bb88d5b/frontera-portal/camino/pprd.cms.settings_local.py#L13-L15)

### CODE REVIEW

File: taccsite_cms/templates/assets_site_delayed.html [hunk 1]
```diff
@@ -12,7 +12,7 @@
 <!-- Site Assets (Delayed): Scripts. -->
 {# FAQ: Bootstrap can use jQuery, Popper, then its own script #}
 {# SEE: https://getbootstrap.com/docs/4.0/getting-started/introduction/#starter-template #}
-<script src="https://code.jquery.com/jquery-3.2.1.slim.min.js" integrity="sha384-KJ3o2DKtIkvYIK3UENzmM7KCkRr/rE9/Qpg6aAZGJwFDMVNA/GpGFF93hXpG5KkN" crossorigin="anonymous"></script>
+<script src="https://code.jquery.com/jquery-3.2.1.min.js" integrity="sha384-xBuQ/xzmlsLoJpyjoggmTEz8OWUFM0/RC5BsqQBDX2v5cMvDHcMakNTNrHIW2I5f" crossorigin="anonymous"></script>
```
Comment (wesleyboar) [inline taccsite_cms/templates/assets_site_delayed.html:15]: [The `slim` jQuery is inadequate for form plugin. I think the size difference is negligible.](https://github.com/TACC/Core-CMS/commit/6834a1800ff778293912b515626677f47a21a036)

<sup>This tiny fix took many hours to figure out. 😭</sup>

File: taccsite_cms/templates/captcha/widget_v2_checkbox.custom.html [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+  data-sitekey="{% for name, value in widget.attrs.items %}{% ifequal name 'data-sitekey' %}{{ value }}{% endifequal %}{% endfor %}"
+  data-callback="onSubmit_{{ widget_uuid }}"
+  data-badge="bottomright"
+  hidden
+>
+  I am not a robot
+</button>
+<script>
+  document.getElementById('tacc-spam-toggle-{{ widget_uuid }}').hidden = false;
+</script>
```
Comment (wesleyboar) [inline taccsite_cms/templates/captcha/widget_v2_checkbox.custom.html:19]: This file shows **how** to create a custom captcha. It has been tested (press a button, captcha passed; javascript off, no captcha button). The file is not loaded. There are instructions (`.md` file of the same name) (next file in diff).

File: taccsite_cms/templates/captcha/widget_v2_checkbox.html [hunk 3]
```diff
@@ -0,0 +1,8 @@
+{# https://github.com/torchbox/django-recaptcha/blob/4921a6c/captcha/templates/captcha/widget_v2_checkbox.html #}
+{% include "captcha/includes/js_v2_checkbox.html" %}
+{# TACC: To change default attr, also add '{% ifnotequal name "data-size" %}' #}
+<div class="g-recaptcha"
+  {% for name, value in widget.attrs.items %}{% if value is not False %} {{ name }}{% if value is not True %}="{{ value|stringformat:'s' }}"{% endif %}{% endif %}{% endfor %}
+>
+</div>
+{# TACC: To customize further, see '../widget_v2_checkbox.custom...' #}
```
Comment (wesleyboar) [inline taccsite_cms/templates/captcha/widget_v2_checkbox.html:8]: The only change to this copied template is adding knowledgeable comments. **Should I delete this file?**

File: taccsite_cms/templates/djangocms_forms/form_template/default.html [hunk 4]
```diff
@@ -0,0 +1,82 @@
+{# https://github.com/avryhof/djangocms-forms/blob/30f4643/djangocms_forms/templates/djangocms_forms/form_template/default.html #}
```
Comment (wesleyboar) [inline taccsite_cms/templates/djangocms_forms/form_template/default.html:1]: The changes to this copied template are commented; the comments begin with `{# TACC: `. The changes are:
1. Load stylesheet ([which could be removed…](…)).
2. Remove redundant script load (fixes negligible console error) (could be done as PR to plugin repo).

**Should I forgo these changes i.e. delete this file?**

**Review by wesleyboar — COMMENTED** [review]
Comments for reviewers.

File: taccsite_cms/settings.py [hunk 5]
```diff
@@ -14,6 +14,11 @@
 import ldap
 from django_auth_ldap.config import LDAPSearch, GroupOfNamesType
 
+from taccsite_cms._settings.form_plugin import *
+from taccsite_cms._settings.form_plugin import (
+    _INSTALLED_APPS as form_plugin_INSTALLED_APPS
+)
```
[the remark this answers is not in the document]
Comment (rstijerina) [reply taccsite_cms/settings.py:20]: I like it 👍

File: taccsite_cms/static/site_cms/css/src/_imports/components/c-button--from-future-core-styles.css [hunk 6]
```diff
@@ -0,0 +1,277 @@
+/* https://github.com/TACC/Core-Styles/blob/d162cfb/source/_imports/components/c-button.css */
```
[the remark this answers is not in the document]
Comment (rstijerina) [reply taccsite_cms/static/site_cms/css/src/_imports/components/c-button--from-future-core-styles.css:1]: What's your plan for integrating this into latest Core-CMS?

File: taccsite_cms/templates/captcha/widget_v2_checkbox.html (hunk 3 above)
Comment (rstijerina) [reply taccsite_cms/templates/captcha/widget_v2_checkbox.html:8]: Up to you.

File: taccsite_cms/static/site_cms/css/src/_imports/components/c-button--from-future-core-styles.css (hunk 6 above)
[the remark this answers is not in the document]
Comment (wesleyboar) [reply taccsite_cms/static/site_cms/css/src/_imports/components/c-button--from-future-core-styles.css:1]: 1. Resolve concerns with this PR.
2. Fix v3.7.0 bugs.
3. Cherry pick from here to update:
    - `task/tup-230-make-form-plugin-work`(which is off of `main`)[^1]
    - i.e. #498
4. Test real good.
5. Request review.

[^1]: This #500 was started by cherry-picking from #498.

File: taccsite_cms/settings.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (taoteg) [reply taccsite_cms/settings.py:20]: concur

File: taccsite_cms/static/site_cms/css/src/_imports/components/c-button.css [hunk 7]
```diff
@@ -0,0 +1,7 @@
+@import url("./c-button--from-future-core-styles.css");
+
+.c-button {
+  --max-width: unset;
+
+  font-size: 1.6rem;
+}
```
[the remark this answers is not in the document]
Comment (taoteg) [reply taccsite_cms/static/site_cms/css/src/_imports/components/c-button.css:7]: _psst: the future is *never* written... ;)_

**Review by taoteg — APPROVED** [review]
Worked for me on frontera following testing steps. LGTM!

**Review by rstijerina — APPROVED** [review]

File: pyproject.toml [hunk 8]
```diff
@@ -48,7 +48,7 @@ djangocms-blog = "1.1.1"
 djangocms-bootstrap4 = "2.0.0"
 djangocms-column = "1.11.0"
 djangocms-file = "3.0.0"
-djangocms-forms = "0.2.5"
+djangocms-forms-maintained = { git = "https://github.com/avryhof/djangocms-forms", rev = "30f464303d317cf7a3170ad64e11ab4471b4287b" }
```
[the remark this answers is not in the document]
Comment (wesleyboar) [reply pyproject.toml:51]: You and I have the same thoughts. I hope a new solution exists by then.

File: taccsite_cms/static/site_cms/css/src/_imports/components/c-button--from-future-core-styles.css (hunk 6 above)
[the remark this answers is not in the document]
Comment (wesleyboar) [reply taccsite_cms/static/site_cms/css/src/_imports/components/c-button--from-future-core-styles.css:1]: Exactly.

If there's an easier route, please tell me it. I have a habit of going the long way unintentionally.

File: taccsite_cms/static/site_cms/css/src/app.forms.css [hunk 9]
```diff
@@ -0,0 +1,5 @@
+/* DO NOT ADD STYLES HERE; ONLY IMPORT OTHER STYLESHEETS */
+
+/* Styles for django cms form plugin */
+@import url("_imports/elements/form.css");
+@import url("_imports/components/django.cms.forms.css");
```
[the remark this answers is not in the document]
Comment (wesleyboar) [reply taccsite_cms/static/site_cms/css/src/app.forms.css:5]: I concur.

<details>

This lets any forms not built by this plugin (i.e. manually built) be already styled. And the "con" of `site.css` I am less bothered by.

</details>

File: taccsite_cms/templates/captcha/widget_v2_checkbox.html (hunk 3 above)
Comment (wesleyboar) [reply taccsite_cms/templates/captcha/widget_v2_checkbox.html:8]: K. I'm gonna keep 'em until I am sure Recaptcha need not be tweaked (e.g. bumped to v3).

**Comment by wesleyboar** [comment]
Process I previously mentioned for deploy was edited; new step to test final UI fix on pre-prod.

File: taccsite_cms/templates/djangocms_forms/form_template/default.html (hunk 4 above)
Comment (wesleyboar) [reply taccsite_cms/templates/djangocms_forms/form_template/default.html:1]: Removed in 8a46a21's update of #498's templates.

1. CSS was moved to `django.cms.forms.css`.
2. PR was created to fix the redundancy in the plugin repo:
  https://github.com/avryhof/djangocms-forms/pull/12

File: taccsite_cms/templates/captcha/widget_v2_checkbox.html (hunk 3 above)
Comment (wesleyboar) [reply taccsite_cms/templates/captcha/widget_v2_checkbox.html:8]: File removed in 8a46a21's update of #498's templates.

File: taccsite_cms/static/site_cms/css/src/_imports/components/c-button--from-future-core-styles.css (hunk 6 above)
[the remark this answers is not in the document]
Comment (wesleyboar) [reply taccsite_cms/static/site_cms/css/src/_imports/components/c-button--from-future-core-styles.css:1]: This files does not exist in #498 (PR for form plugin off of `main`). [I load from core-styles.](https://github.com/TACC/Core-CMS/commit/d813bee5f15c07474e6a1971debfc742c38165b0#diff-78264304a022041823ede0e08af9cca803a7118c24d4466bf8b44deac1ebf58b)

**Pull request merged**